File: src/SolverExact.cpp


#include "../include/SolverExact.h"


// constructors and destructors
SolverExact::SolverExact(double suboptimal_threshold, bool silent) : Solver(silent)
{
    suboptimal_threshold_ = suboptimal_threshold;
}

SolverExact::~SolverExact() { }


double find_smallest_distance_range(int self_index,
                                    RangeIndex range,
                                    std::vector< std::vector<double> > distance_matrix)
{
    // find closest elements in other range and the distance
    double best_distance = std::numeric_limits<double>::infinity();
    double distance;

    for (int i = range.low; i != range.high; ++i)
    {
        distance = distance_matrix[self_index][i];
        if (distance < best_distance)
        {
            best_distance = distance;
        }
    }
    return best_distance;
}


std::vector< std::vector<double> > elem_range_distance_matrix(std::vector<RangeIndex> ranges,
                                                              std::vector< std::vector<double> > distance_matrix)
{
    // using the distance matrix, find closest neighbors for
    // each object, in each other groups (range)

    // matrix is (distance_matrix.size() x ranges.size()), initialize it
    std::vector< std::vector<double> > elem_range_dm = std::vector< std::vector<double> >();
    for(size_t i = 0; i != distance_matrix.size(); ++i)
    {
        elem_range_dm.push_back(std::vector<double>(ranges.size(), 0));
    }

    // for each elem, for each range, find the closest distance between the two
    for (size_t elem_index = 0; elem_index != distance_matrix.size(); ++ elem_index)
    {
        for (size_t range_index = 0; range_index != ranges.size(); ++ range_index)
        {
            elem_range_dm[elem_index][range_index] = find_smallest_distance_range(elem_index,
                                                                                  ranges[range_index],
                                                                                  distance_matrix);
        }
    }
    return elem_range_dm;
}


bool is_leaf(SubSolution sol)
{
    // checks wether or not a solution is complete
    return (sol.get_ranges().size() == 0);
}


void expand_solution(SubSolution solution,
                     std::vector< std::vector<double> > elem_range_dm,
                     std::vector< std::vector<double> > distance_matrix,
                     std::priority_queue< SubSolution > & search_space,
                     double cutoff)
{
    // used to create all children of a node
    // and add them, if they are promising enough, to the
    // search space (which is a max heap)
    assert((is_leaf(solution) == false) && "applying on complete solution");

    // separate the ranges, the one to explore and the rest
    RangeIndex range_to_explore = solution.get_ranges()[0];
    std::vector<RangeIndex> new_ranges = std::vector<RangeIndex>();
    for (size_t i = 1; i != solution.get_ranges().size(); ++i)
    {
        new_ranges.push_back(solution.get_ranges()[i]);
    }

    for (size_t new_gene = range_to_explore.low;
         new_gene != range_to_explore.high;
         ++new_gene)
    {
        // create the new genes
        std::vector<int> new_genes = std::vector<int>();
        for (size_t g = 0; g != solution.get_genes().size(); ++g)
        {
            new_genes.push_back(solution.get_genes()[g]);
        }
        new_genes.push_back(new_gene);

        // create the new solution, score it and add it if its worth it
        SubSolution new_solution = SubSolution(new_genes, new_ranges);
        new_solution.set_score(elem_range_dm, distance_matrix);

        if (new_solution.get_score() <= cutoff)
        {
            search_space.push(new_solution);
        }
    }
    return;
}


bool compare_range(RangeIndex A, RangeIndex B) { return (A.high - A.low) < (B.high - B.low); }


std::vector<Solution> SolverExact::solve(std::vector< std::vector<double> > distance_matrix,
                                         std::vector<Range> ranges) const
{
    // use a priority queue to represent the current search space
    std::priority_queue<SubSolution> search_space = std::priority_queue<SubSolution>();
    std::vector<SubSolution> satisfying_solutions = std::vector<SubSolution>();
    double best_score = std::numeric_limits<double>::max();
    double scaled_threshold = suboptimal_threshold_ * (ranges.size() * (ranges.size()-1));

    // sort the ranges by their size (to avoid big branching at the start)
    std::vector<RangeIndex> sorted_ranges = std::vector<RangeIndex>();
    for (int i = 0; i != ranges.size(); ++i)
    {
        RangeIndex converted = {i, ranges[i].low, ranges[i].high};
        sorted_ranges.push_back(converted);
    }
    std::sort(sorted_ranges.begin(), sorted_ranges.end(), compare_range);


    // initialize the search space with the empty subsolution
    search_space.push(SubSolution(std::vector<int>(), sorted_ranges));
    std::vector< std::vector<double> > e_r_dm = elem_range_distance_matrix(sorted_ranges, distance_matrix);


    int iteration = 0;
    while (search_space.size() > 0)
    {

        // fetch current solution
        SubSolution current_sol = search_space.top();
        search_space.pop();

        // if full solution, check if good enough and add to the best or delete it
        if (is_leaf(current_sol))
        {
            // check if worth adding to satisfying solutions
            if ((current_sol.get_score() <= best_score + scaled_threshold))
            {
                satisfying_solutions.push_back(current_sol);
            }
            // check if the score was actually the best seen and update if it is
            if (current_sol.get_score() < best_score)
            {
                best_score = current_sol.get_score();
            }
        } else { // not a complete solution
            expand_solution(current_sol, e_r_dm, distance_matrix, search_space, best_score + scaled_threshold);
        }

        iteration += 1;
    }

    // filter out the solutions
    std::vector<Solution> final_solutions = std::vector<Solution>();
    for (size_t i = 0; i != satisfying_solutions.size(); ++i)
    {
        // add only those satisfying scoring constraints
        if (satisfying_solutions[i].get_score() <= best_score + scaled_threshold)
        {
            final_solutions.push_back(satisfying_solutions[i].return_solution());
        }
    }
    return final_solutions;
}

File: src/src/mccons2/optimization/SolverExact.java
package mccons2.optimization;

import mccons2.util.Pair;
import mccons2.util.WidthComparator;

import java.util.ArrayList;
import java.util.Collections;
import java.util.PriorityQueue;



class SubSolution extends Solution{


    private ArrayList<Pair<Integer, Integer>> ranges;


    public SubSolution(ArrayList<Integer> genes_,
                       double score_,
                       ArrayList<Pair<Integer, Integer>> ranges_)
    {
        super(genes_, score_);
        ranges = ranges_;
    }


    public ArrayList<Pair<Integer, Integer>> getRanges() {
        return ranges;
    }


    public String toString()
    {
        // add the genes
        StringBuilder builder = new StringBuilder();
        builder.append("genes ");
        for (Integer i : getGenes())
        {
            builder.append(i + " ");
        }
        builder.append(System.lineSeparator());

        // add the score
        builder.append("score " + getScore() + System.lineSeparator());

        // add the remaining ranges
        builder.append("ranges" + System.lineSeparator());
        for( Pair<Integer, Integer> r : getRanges())
        {
            builder.append("[ " + r.getFirst() + " .. " + r.getSecond() + " ]");
            builder.append(System.lineSeparator());
        }
        return builder.toString();
    }


    public Solution toSolution()
    {
        return new Solution(getGenes(), getScore());
    }
}




public class SolverExact extends Solver {


    private double threshold;
    private boolean verbose;

    public boolean isVerbose()
    {
        return verbose;
    }

    public SolverExact(double suboptimal_threshold, boolean verbose_)
    {
        threshold = suboptimal_threshold;
        verbose = verbose_;
    }


    /**
     * find closest elements in other range and the get
     * @param self_index
     * @param range
     * @param distanceMatrix
     * @return
     */
    public static double findSmallestDistanceRange(int self_index, Pair<Integer, Integer> range,
                                                   double[][] distanceMatrix) {

        double best_distance = Double.POSITIVE_INFINITY;
        double distance;

        for (int i = range.getFirst(); i != range.getSecond(); ++i) {
            distance = distanceMatrix[self_index][i];
            if (distance < best_distance) {
                best_distance = distance;
            }
        }
        return best_distance;
    }


    /**
     * using the get matrix, find closest neighbors for
     * each object, in each other groups (range)
     * @param ranges
     * @param distanceMatrix
     * @return
     */
    double[][] nearestNeighborDistanceElemToRange(ArrayList<Pair<Integer, Integer>> ranges,
                                                  double[][] distanceMatrix) {

        // matrix is (distanceMatrix.size() x ranges.size())
        double[][] elem_range_dm = new double[distanceMatrix.length][ranges.size()];

        // for each elem, for each range, find the closest get between the two
        for (int elem_index = 0; elem_index != distanceMatrix.length; ++elem_index) {
            for (int range_index = 0; range_index != ranges.size(); ++range_index) {
                elem_range_dm[elem_index][range_index] =
                        findSmallestDistanceRange(elem_index, ranges.get(range_index), distanceMatrix);
            }
        }
        return elem_range_dm;
    }


    boolean isCompleteSolution(SubSolution sol) {
        // checks whether or not a solution is complete
        return (sol.getRanges().size() == 0);
    }


    void branch(SubSolution solution,
                double[][] elem_range_dm,
                double[][] distanceMatrix,
                PriorityQueue<SubSolution> search_space,
                double cutoff) {
        // used to create all children of a node
        // and add them, if they are promising enough, to the
        // search space (which is a max heap)
        assert (isCompleteSolution(solution) == false);

        // separate the ranges, the one to explore and the rest
        Pair<Integer, Integer> range_to_explore = solution.getRanges().get(0);
        ArrayList<Pair<Integer, Integer>> new_ranges = new ArrayList<>();

        for (int i = 1; i != solution.getRanges().size(); ++i) {
            new_ranges.add(solution.getRanges().get(i));
        }

        for (int new_gene = range_to_explore.getFirst();
             new_gene != range_to_explore.getSecond();
             ++new_gene) {

            // create the new genes
            ArrayList<Integer> new_genes = new ArrayList<>();
            for (int g = 0; g != solution.getGenes().size(); ++g) {
                new_genes.add(solution.getGenes().get(g));
            }
            new_genes.add(new_gene);

            // create the new solution, score it and add it if its worth it
            SubSolution new_solution = new SubSolution(new_genes, Double.POSITIVE_INFINITY, new_ranges);
            assignPairwiseDistanceScore(new_solution, distanceMatrix);

            if (new_solution.getScore() <= cutoff) {
                search_space.add(new_solution);
            }
        }
    }


    public ArrayList<Solution> solve(double[][] distanceMatrix, ArrayList<Pair<Integer, Integer>> ranges)
    {
        // use a priority queue to represent the current search space
        PriorityQueue<SubSolution> searchSpace = new PriorityQueue<>();
        ArrayList<SubSolution > satisfyingSolutions = new ArrayList<>();
        double bestScore = Double.POSITIVE_INFINITY;
        double leeway = threshold * (ranges.size() * (ranges.size() - 1));
        double scoreThreshold = bestScore + leeway;

        // sort the ranges by their size (to avoid big branching at the start)
        ArrayList<Pair<Integer, Integer> > sortedRanges = new ArrayList<>(ranges);
        Collections.sort(sortedRanges, new WidthComparator());


        // initialize the search space with the empty subsolution
        searchSpace.add(new SubSolution(new ArrayList<Integer>(), Double.POSITIVE_INFINITY, sortedRanges));
        double[][] elemToRangeDistanceMatrix = nearestNeighborDistanceElemToRange(sortedRanges, distanceMatrix);


        // Branch & Bound main loop
        while (searchSpace.size() > 0) {
            scoreThreshold = bestScore + leeway;
            SubSolution currentSolution = searchSpace.poll();

            if (isCompleteSolution(currentSolution)) {
                // check if worth adding to satisfying solutions
                if ((currentSolution.getScore() <= scoreThreshold)) {
                    satisfyingSolutions.add(currentSolution);
                }
                // check if the score was actually the best seen and update if it is
                if (currentSolution.getScore() < bestScore) {
                    bestScore = currentSolution.getScore();
                }
            } else {  // not a complete solution
                branch(currentSolution, elemToRangeDistanceMatrix, distanceMatrix, searchSpace, scoreThreshold);
            }
        }

        // filter out solutions by score knowing now the best possible score
        ArrayList<Solution > bestSolutions = new ArrayList<>();
        for (int i = 0; i != satisfyingSolutions.size(); ++i) {
            // add only those satisfying scoring constraints
            if (satisfyingSolutions.get(i).getScore() <= scoreThreshold) {
                bestSolutions.add(satisfyingSolutions.get(i).toSolution());
            }
        }
        return bestSolutions;
    }
}